esrally/driver/runner.py

```diff
@@ -891,6 +910,114 @@ async def _search_after_query(es, params):
+                "timed_out": False,
+                "took": 0,
+            }
+            if pit_op:
+                # these are disallowed as they are encoded in the pit_id
+                for item in ["index", "routing", "preference"]:
+                    body.pop(item, None)
+                index = None
+            # explicitly convert to int to provoke an error otherwise
+            total_pages = sys.maxsize if params.get("pages", "all") == "all" else int(mandatory(params, "pages", self))
+            for page in range(1, total_pages + 1):
+                if pit_op:
+                    pit_id = CompositeContext.get(pit_op)
+                    body["pit"] = {"id": pit_id, "keep_alive": "1m"}
+
+                paths_to_composite = paths_to_composite_agg(body, [])
+                if not paths_to_composite or len(paths_to_composite) != 1:
+                    raise Exception("Unique path to composite agg required")
+                path_to_composite = paths_to_composite[0]
+                composite_agg_body = resolve_composite_agg(body, path_to_composite)
+                if not composite_agg_body:
+                    raise Exception("Could not find composite agg - parser inconsistency")
+                if size:
+                    composite_agg_body["size"] = size
+
+                body_to_send = tree_copy_composite_agg(body, path_to_composite)
+                response = await self._raw_search(es, doc_type=None, index=index, body=body_to_send, params=request_params, headers=headers)
+                parsed = self._composite_agg_extractor(response, bool(pit_op), path_to_composite, results.get("hits"))
+                results["pages"] = page
+                results["weight"] = page
+                if results.get("hits") is None:
+                    results["hits"] = parsed.get("hits.total.value")
+                    results["hits_relation"] = parsed.get("hits.total.relation")
+                results["took"] += parsed.get("took")
+                # when this evaluates to True, keep it for the final result
+                if not results["timed_out"]:
+                    results["timed_out"] = parsed.get("timed_out")
+                if pit_op:
+                    # per the documentation the response pit id is most up-to-date
+                    CompositeContext.put(pit_op, parsed.get("pit_id"))
+
+                after_key = parsed["after_key"]
+                self.logger.info("Afer key is [%s]", after_key)
+                if isinstance(after_key, dict):
+                    composite_agg_body["after"] = after_key
+                else:
+                    # body needs to be un-mutated for the next iteration (preferring to do this over a deepcopy at the start)
+                    body.pop("pit", None)
+                    composite_agg_body.pop("after", None)
+                    break
+
+            return results
+
+        def select_aggs(obj):
+            if isinstance(obj, dict):
```

Comment:
I think we can refactor the `isinstance` checks in this function (as well as in `paths_to_composite_agg`, `resolve_composite_agg`, and `tree_copy_composite_agg` below.)  As far as I can tell--and let me know if I'm missing something--everywhere `isinstance` is called in these functions, the variable whose type is being inspected can only ever be a `dict` or `None`. 

This is because `body` is what's being passed through the chain of function calls, and it can only ever be a `dict`. Along the way, we can also get intermediate `None` return values. I think that covers all cases, unless I'm misreading.

If so, this function could look something like:

```python
def select_aggs(obj):
    if obj is not None:
        return obj.get("aggs", None) or obj.get("aggregations", None)
    else:
        return None
```

Reply:
There's two types of functions here. The first ones are `paths_to_composite_agg` (and the `select_aggs` function that is being used within), which figures out whether the given user-defined request body contains a composite agg definition. The parsing here cannot assume for the request body to be well-formed, and is rather lenient in the parsing (i.e. double-checks that we have dicts in the right place) so that it can later throw a proper error stating that it couldn't find the composite agg. If we were to just assume that some of these definitions would be dicts, we would give a cryptic Python error message accessing a non-dict in the wrong way.

For the above I prefer the current parsing code.

There are two more functions, `resolve_composite_agg` and `tree_copy_composite_agg`, where we use as input whatever was derived via `paths_to_composite_agg`, and hence we can assume that the right structure and types are in place. I made the code in those functions [less defensive](https://en.wikipedia.org/wiki/Defensive_programming), as `paths_to_composite_agg` has already checked the right shape of dicts.

I hope this addresses your concerns.

Reply:
Sure, this makes sense. Thanks!